Add CompOSE EOS table reader

nilsdeppe opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add CompOSE EOS table reader

kidder commented on src/IO/ComposeTable.cpp:
```diff
@@ -0,0 +1,224 @@
+  std::string line_buffer{};
+  const std::string filename{directory_to_read_from_ + "/eos.quantities"};
+  if (not file_system::check_if_file_exists(filename)) {
+    ERROR("File '" << filename << "' does not exist.");
```
test your ERRORs or exclude them from coverage

kidder reviewed (changes requested):
squash

nilsdeppe committed:
Add copy to file_system

nilsdeppe committed:
Alphabetize FileSystem.?pp

nilsdeppe committed:
Refactor file system error tests

nilsdeppe committed:
Switch most of FileSystem to std::filesystem

The remaining ones are:
- get_parent_path
- glob

These don't trivially translate into std::filesystem but shouldn't be too
difficult to switch with some effort.

nilsdeppe commented:
Added tests for the errors. I needed `file_system::copy` for this and then changed most of our file system library to use `std::filesystem`. This means there's less we need to maintain and should make us more resilient to OS, compiler, etc. switches

kidder commented on src/Utilities/FileSystem.cpp:
```diff
@@ -67,43 +69,21 @@ void create_directory(const std::string& dir, const double wait_time,
           wait_time *
           std::pow(wait_time_increase_factor, number_of_failures + 1)));
-      // LCOV_EXCL_STOP
+      // LCOV_EXCL_STOR
```
?

kidder reviewed (changes requested):
the failing test does do file manipulation so you need to investigate...

nilsdeppe committed:
Drop support for clang-9

std::filesystem is completely incompatible with clang-9 in the container and
using std::filesystem simplifies a lot of our filesystem library.
